R/finalize_plot.R

```diff
@@ -315,29 +338,113 @@ finalize_plot <- function(input_plot = NULL,
+
+    # Add left indent to plot margins
+    updated_l_margins <- c(l_margins[1:3],l_margins[4] + l_indent)
+
+    # Create a ggplot table of the updated plot element, including a left-aligned legend
+    gt <- ggplot2::ggplot_gtable(
+      ggplot2::ggplot_build(p + theme(legend.position = "left",
+                                      legend.justification = "left",
+                                      legend.direction = "horizontal",
+                                      plot.title = element_blank(),
+                                      plot.margin = grid::unit(p_margins,"bigpts"),
+                                      legend.margin = margin(t = updated_l_margins[1],
+                                                             r = updated_l_margins[2],
+                                                             b = updated_l_margins[3],
+                                                             l = updated_l_margins[4],
+                                                             "bigpts")
+                                      )
+                            )
+      )
+
+    # Extract the legend
+    l <- gtable::gtable_filter(gt, "guide-box")[[1]]
+
+    # Remove the legend from the plot
+    p <- p + ggplot2::theme(legend.position = "none")
+
+    # Create a buffer
+    buffer <- grid::rectGrob(gp = grid::gpar(col = "transparent", fill = "transparent"))
+
+    # Assemble
+    built <- gridExtra::grid.arrange(
+      gridExtra::arrangeGrob(
+        l[[1]],
+        buffer,
+        p,
+        nrow = 3,
+        heights = grid::unit(c(l_height,
+                               plot_constants$margin_v5,
+                               p_height),
+                             "bigpts")
+      )
+    )
+
+    return(built)
+  }
+
+  # Use helper function to develop full stack of legend, buffer, and plot
+  p_l_stack <- buildChart(processed_plot,
+                          l_margins = plot_constants$padding_legend,
+                          p_margins = plot_constants$padding_plot,
+                          l_indent = plot_constants$legend_indent,
+                          l_height = legend_height,
+                          p_height = plot_constants$height - legend_height - plot_constants$margin_v1_v2_v4_v5)
+
   # ggplot as grob (vp.plot)
   grob_plot <- grid::grobTree(
-    ggplotGrob(
-      input_plot + ggplot2::theme(
-        # make sure the plot has no title or caption
-        plot.title = ggplot2::element_blank(),
-        plot.caption = ggplot2::element_blank(),
-        # add margins
-        plot.margin = unit(plot_constants$padding_plot, "bigpts"),
-        # modify text sizing. **THIS MODIFICATION IS NEEDED BUT NOT UNDERSTOOD**
-        text = ggplot2::element_text(size = cmapplot_globals$font$main$size * 1.25)
-      )
-    ),
+    p_l_stack,
     vp = vp.plot,
     name = "plot"
   )
 
+  # If debug is called, assemble additional outlines for overlay
```

Comment:
I'd lean towards drawing outlines around the actual objects, as opposed to assembling additional objects that serve as outlines. In a buggy environment, there's a chance that your debug_stack won't actually align with where grobs and viewports are finding themselves placed.

Reply:
If you know a good way to do this, that would be great. I had a great deal of trouble trying to add outlines to the objects in the ggtable, but that could just be unfamiliarity on my part. It didn't want me to add boundaries. Also, you can't add a border to a ggplot generally - it just lets you add a border around the panel, but not the labels.

Reply:
Not off the top of my head but I'd be happy to look into it later this week.

Reply:
I tried adding a border again with the new method, but it remains stubbornly not showing up. I've left the overlays in for now, as I don't have any better ideas at the moment.

Reply:
I've addressed this via borders in the ggplot theme.